### ISSUE TITLE
Add Curtain devices

### ISSUE THREAD

**Issue by tonysprenk** [basdelfos/homebridge-tuya-web#9]
Hi, any plans to add curtains to the plugin? I  would be willing to test, I have 5 sets of curtains that I would like to add to the home app. Thanks!

**Comment by basdelfos** [comment]
I already did some coding for a previous request, but I don’t have any curtain devices to try out. Do you have coding experience or are you able to do some tests with Postman for me?

**Comment by tonysprenk** [comment]
Yes we spoke over email about the curtains before, I just thought for the sake of transparency in would also open an issue. I have very minimal coding experience and I have never used postman before, but I can usually figure things out pretty quickly.

**Comment by jbiscarri** [comment]
Also interested in courtains & shutters. If I can help yoy, please, let me know how. Thanks

**Comment by tonysprenk** [comment]
Any update on this?

**Comment by PierreBrisorgueil** [comment]
interested too :) 

I've tried today : 
```
[TuyaWebPlatform] Adding: Curtain switch (cover / xxxxx)
[TuyaWebPlatform] Could not init class for device type [cover]
[TuyaWebPlatform] Could not find accessory in dictionary
```

**Comment by fraho67** [comment]
Bump
Had the same result as PierreBrisorgueil yesterday, Thought it would work as these curtain devices are just switches with on/off functionality.

**Comment by basdelfos** [comment]
As I don't own such a device it is really hard to test and debug to get this to work.
If someone with coding skills and owns such a device can contribute to this repository, that would be great. Until that moment adding this feature is on hold.

**Comment by fraho67** [comment]
I'm an IT guy but not in coding. I could contribute in testing an help you with debugging

**Comment by fraho67** [comment]
> > @jbiscarri
> > But what about the key? If it would be the Id only I had done this a long time ago
> 
> with TuyaLan method you need the id and the key for every device. You can get it following this steps: https://github.com/AMoo-Miki/homebridge-tuya-lan/wiki/Setup-Instructions

Thanks, with this method I managed to get the key - unfortunately TuyaLan now gives me the message "failed to discover <device> in time

**Comment by mladjan** [comment]
Hi guys,

I have curtain switch and have coding experience. How I can help to add support for Curtains? :)

**Comment by fraho67** [comment]
@mladjan 

I guess @basdelfos  meant that someone with coding experience should look at the source code trying to figure out how these curtain devices can be implemented

**Comment by sidbmx** [comment]
I’d love @mladjan to add the homekit support for the curtains/blinds for the Tuya-homebridge !
I have 4 shutters now controlled in Tuya. Rest of the house is homekit.

**Comment by costispavlou** [comment]
I'm using the curtains/blinds/shutter switches from Tuya as dimmers. They work, but if you could natively support for them to be shown as blinds I'd very much appreciate it.

**Comment by AndreyBelyakov** [comment]
@costispavlou how do you use your dimmers: doest is correspond from 0(closed) to 100 (open)? for me only opens the curtain when it is light or dimmer, but can not control the position

**Comment by costispavlou** [comment]
My shutter switches have 3 positions : up / stop / down.
In the homekit.app they show as lights. ON / OFF. pressing on they open all the way (and don't stop until a couple of seconds after they are open) and pressing off they close down (and again dont stop until a couple of seconds after they are closed). 

They work as they suppose to minus the pause/stop button and the visual interface of 'blinds' instead of lights. Therefore asking Siri to turn off the office lights, will close the shutters too.

**Comment by vitramir** [comment]
@AndreyBelyakov @costispavlou you can check my branch 
https://github.com/vitramir/homebridge-tuya-web/tree/feature/window-covering

I use native HomeApp Window Covering service and override step to 50. So I get dimmer UI with 3 states - 0/50/100. 

https://github.com/vitramir/homebridge-tuya-web/blob/feature/window-covering/src/window_covering.ts#L86

**Comment by costispavlou** [comment]
> @AndreyBelyakov @costispavlou you can check my branch
> https://github.com/vitramir/homebridge-tuya-web/tree/feature/window-covering
> 
> I use native HomeApp Window Covering service and override step to 50. So I get dimmer UI with 3 states - 0/50/100.
> 
> https://github.com/vitramir/homebridge-tuya-web/blob/feature/window-covering/src/window_covering.ts#L86

I'm kinda lost. 
This is my current config.
let me know what I'm doing wrong. Should I install anything?

{
    "platform": "TuyaWebPlatform",
    "name": "TuyaWebPlatform",
    "options": {
        "username": "************@*******.com",
        "password": "********************",
        "countryCode": "357",
        "platform": "tuya",
        "pollingInterval": 10
    },
    "defaults": [
        {
            "id": "****************",
            "device_type": "dimmer"
        },
        {
            "id": "60602070************",
            "device_type": "dimmer"
        }
    ]
}

**Comment by vitramir** [comment]
@costispavlou 
If you want to try my branch you should replace package `homebridge-tuya-web` with `@vitramir/homebridge-tuya-web`.  Please note that only the latest version of the homebridge supports scoped packages (like @USERNAME/PACKAGE_NAME)

After you install my version of the package, you should set device_type to `cover`

And I noticed an issue with changing device_type for previously connected devices. May be you have to remove ~/.homebridge/accessories/cachedAccessories

**Comment by costispavlou** [comment]
so after I remove tuya-web I need to install @vitramir/homebridge-tuya-web ?

**Comment by vitramir** [comment]
Yep. I think it is exactly what "replace" mean :)

**Comment by costispavlou** [comment]
thank you so much. How can I change the timing of the covers ?
They are stuck in 'closing'

**Comment by vitramir** [comment]
Looks like this is a new enhancement. My switch automatically switches to stop/pause after couple of minutes and it is ok for me. We can add timer and change the switch state to stop/pause after configurable delay.

**Comment by costispavlou** [comment]
screenshots. 

![0-02-05-3cba524c8d80e0a0ac3da5f5ed418668f24f9de44072ea55b51a41931609e253_c97ba02c](https://user-images.githubusercontent.com/57916603/78355866-2d6ffb80-75b7-11ea-97c9-838d186dd46a.jpg)

after a couple of minutes of the above, they show as 50% 

![0-02-05-3cba524c8d80e0a0ac3da5f5ed418668f24f9de44072ea55b51a41931609e253_c97ba02c](https://user-images.githubusercontent.com/57916603/78355832-15987780-75b7-11ea-8263-3165167793aa.jpg)

**Comment by vitramir** [comment]
Yes. I don't like this too, but I don't understand how to solve this without a position sensor.

We have only 3 states. opening/stopped/closing. We can use cache, write some logic at our homebridge side and extend states. For example `stopped` after `closing` should become `closed`. But if you manually change switch state, then synchronization will be lost and displayed state in HomeApp will be wrong.

There is another possible solution. I want to try using Tuya OpenAPI. (https://github.com/basdelfos/homebridge-tuya-web/issues/82)
It allows communication with devices over the lan and it will give us very very low ping. So we will catch manual changes of the switch state and track curtains position. (With configured open/close time in seconds of course). Theoretically is will be possible to set curtains to exact %.

**Comment by costispavlou** [comment]
can you set it as Open (active) / 50% (active / Closed (inactive) ?

**Comment by vitramir** [comment]
Not sure what you mean. 

There is only one state parameter in Switch - 1/2/3

There are three characteristics for this accessory type in HomeApp:
TargetPosition - from 0 to 100%
```
1->100
2->0
3->50
```

CurrentPosition - from 0 to 100%
```
1->100
2->0
3->50
```

PositionState - INCREASING, DECREASING, STOPPED
```
1->INCREASING
2->DECREASING
3->STOPPED
```

I can move this mapping to config. But how do you want to change it?

**Comment by vitramir** [comment]
To make it really nice we should track position. Get it from the sensor or calculate. To calculate we need the delay between physically changing switch state and callback in our code to be as low as possible. 
So the easy way is - OpenAPI.
The hard way is - custom firmware.

I don't know others.

**Comment by costispavlou** [comment]
The opening state should remain as is, but at the end return OPEN and not revert to 50%.
The closing state should be changed to return CLOSED and set to (inactive). Again not revert to 50%. 
So i guess change it as such??
TargetPosition - from 0 to 100%
1->100
2->0
3->50%
CurrentPosition - from 0 to 100%
1->100
2->Closed
3->Open

PositionState - INCREASING, DECREASING, STOPPED
1->Opening
2->Closing
3->Open

**Comment by costispavlou** [comment]
actually the way it can be seen as it was when in dimmer mode was perfect. 
The only thing i can think of is 0-100 should be whatever if in between, whereas 0 = closed and 100 = open

**Comment by vitramir** [comment]
First case:
You are closing your curtains. After couple of minutes your switch will change to State=3. This will change PositionState=Open.

Second case:
You manually close curtains and press stop button. Homebridge has no information about this action (because state was 3 at previous update and still 3 again) and shows you that window is Open.

How do you want to solve it?

**Comment by vitramir** [comment]
Current dimmer device works based on cache. If you manually close curtains and press stop faster then polling interval (10s by default). You will get wrong state in HomeApp

**Comment by costispavlou** [comment]
> First case:
> You are closing your curtains. After couple of minutes your switch will change to State=3. This will change PositionState=Open.
> 
> Second case:
> You manually close curtains and press stop button. Homebridge has no information about this action (because state was 3 at previous update and still 3 again) and shows you that window is Open.
> 
> How do you want to solve it?

Actually it defaults to 50% not open/closed. 
I only need an Open / Closed state to be honest

**Comment by vitramir** [comment]
I can add cache and this logic
opening + stopped => open
closing + stopped => closed

But right after you stop it manually (set to 50%) it will automatically change it's state to open/closed.

**Comment by costispavlou** [comment]
sure let's try it out why not.

**Comment by costispavlou** [comment]
@vitramir did you manage to come up with a way not showing the blinds 50% open?
I'm here to try it out if you want!

**Comment by vitramir** [comment]
I haven’t had time to do this yet.

**Comment by costispavlou** [comment]
can I help you in anyway? 
It's just that, right now everything defaults to/shows as 50% open, despite being closed or open.

**Comment by vitramir** [comment]
You can add cache and some logic by yourself of course.

**Comment by costispavlou** [comment]
if I only knew how hehe :) 
can you pinpoint me somewhere? I use hoobs btw.

**Comment by vitramir** [comment]
If you have no experience with nodejs, typescript, promises and etc I think I may be difficult for you. You can just wait a little bit :)

If you still want to try then:
This is the place where you can catch changes and update device state https://github.com/vitramir/homebridge-tuya-web/blob/feature/window-covering/src/window_covering.ts#L86

This is an example of cache:
Get:
https://github.com/vitramir/homebridge-tuya-web/blob/feature/window-covering/src/dimmer_accessory.ts#L138

Set:
https://github.com/vitramir/homebridge-tuya-web/blob/feature/window-covering/src/dimmer_accessory.ts#L193

After updating sources you have to run `npm run tsc` or 'yarn tsc' to compile.

**Comment by costispavlou** [comment]
ok i'll wait for when you have the time then :)

**Comment by vitramir** [comment]
@costispavlou you can try it

```@vitramir/homebridge-tuya-web@0.1.10```

https://github.com/vitramir/homebridge-tuya-web/commit/3401dbfb1ef6d006517c73f0e09a3e6c96544d35

**Comment by costispavlou** [comment]
it works great!
thank you so much

**Comment by fraho67** [comment]
@costispavlou 
May I ask you what kind of blindswitches you are using? I've also tried your workaround by defining mine as dimmers, but no luck.
Mine are these ones https://www.amazon.de/gp/product/B07PS5LRPQ/ref=ppx_yo_dt_b_asin_title_o09_s00?ie=UTF8&th=1

**Comment by vitramir** [comment]
I use this one https://www.aliexpress.com/item/32972956446.html

**Comment by costispavlou** [comment]
> @costispavlou
> May I ask you what kind of blindswitches you are using? I've also tried your workaround by defining mine as dimmers, but no luck.
> Mine are these ones https://www.amazon.de/gp/product/B07PS5LRPQ/ref=ppx_yo_dt_b_asin_title_o09_s00?ie=UTF8&th=1

hey man, 
seriously try vitramirs plugin and you'll be pleased. 
I use these. 
https://www.aliexpress.com/item/Tuya-Smart-Life-WiFi-Curtain-Blind-Switch-for-Roller-Shutter-Electric-motor-Google-Home-Alexa-Echo/32963064005.html?spm=a2g0s.9042311.0.0.27424c4dcSMRCU

**Comment by fraho67** [comment]
Thanks for sharing your devices, mine however still do not work, neither with vitramirs plugin nor with tuya-lan

**Comment by costispavlou** [comment]
> Thanks for sharing your devices, mine however still do not work, neither with vitramirs plugin nor with tuya-lan

first of all, did you set them up correctly? 
do they work with the tuya app ?
if yes, then the issue is with your config file. 
let me help you. copy paste the config (obviously remove your password)

**Comment by fraho67** [comment]
Yes, it does work with the Tuya-App. My (Tuyaweb-)config is this one. I've also removed the id. I have two other devices but tested the workaround with only one of it.

{
            "platform": "TuyaWebPlatform",
            "name": "TuyaWebPlatform",
            "options": {
                "username": "*******",
                "password": "******",
                "countryCode": "49",
                "platform": "tuya",
                "pollingInterval": 10
            },
            "defaults": [
                {
                    "id": "*******",
                    "device_type": "dimmer"
                }
            ]
        },

**Comment by costispavlou** [comment]
remove that plugin and install the one vitramir/homebridge-tuya-web@0.1.10
then your config should look like this:

{
    "platform": "TuyaWebPlatform",
    "name": "TuyaWebPlatform",
    "options": {
        "username": "*",
        "password": "*",
        "countryCode": "*",
        "platform": "tuya",
        "pollingInterval": 10
    },
    "defaults": [
        {
            "id": "*",
            "device_type": "cover"
        },
        {
            "id": "*",
            "device_type": "cover"
        }
    ]
}

**Comment by fraho67** [comment]
Yesterday I had vitramirs 0.1.11 installed but I will give it another try as soon as I have some time, thanks

**Comment by costispavlou** [comment]
you are correct. the latest should be 0.1.11
just change the types to cover and try again

**Comment by fraho67** [comment]
So I've tried it changing the type to "cover" with the same result (devices not responding/working in Homekit) What's strange is, that whenever I'm doing sudo rm -rf accessories within Homebridge none of the devices vanishes as if I hadn't deleted anything

**Comment by fraho67** [comment]
Meanwhile I've noticed that I do not have to change the type to "cover" or anything else. It's working right away. This issue with the status however is that if I tap on an icon for opening it says "opening" and it's keeping that status (same with closing), it's not changing the status after a couple of minutes to "opened". Was that the reason for the "cover"-thing? Does that change the behaviour?

**Comment by costispavlou** [comment]
cover thing makes the blinds to be seen as blinds instead of dimmers (lights)

**Comment by fraho67** [comment]
Changed one of the blinds back to "cover" but it seems that it's the same with the status. But maybe this is some kind of translation error. In german it's changing the status from "Wird geöffnet" to "Öffnen" which is in english "will be opened" and "open" . The last one should be in german "Offen" or "geöffnet". For a blind that has the english status "closed" it would be "geschlossen". Is this something that can be manipulated via the plugin? @costispavlou I'm getting the same longpress-menu for all of my three blinds, no matter if it's changed to "cover" or not

**Comment by costispavlou** [comment]
hi, since the plugin is not mine i'm not able to tell if that plays any part in this. 
I have my language set to english and mine works perfectly. did not need to change anything else other than blinds / cover

**Comment by fraho67** [comment]
It's odd....the home app on my Macbook says "geöffnet" (which is correct) but the home app on my iPhone it says "öffnen" (which is not correct). So this has imo nothing to do with the plugin
